pkg/decoder/delimiter.go

```diff
@@ -0,0 +1,145 @@
+	}
+}
+
+// Delimiter
+type Delimiter interface {
+	LineChan() chan *Line
+}
+
+// SingleLineDelimiter reads lines from lineChan and uses lineBuffer to send them
+type SingleLineDelimiter struct {
+	lineChan   chan *Line
+	lineBuffer *LineBuffer
+}
+
+// NewSingleLineDelimiter returns a new NewSingleLineDelimiter
+func NewSingleLineDelimiter(outputChan chan *Output) *SingleLineDelimiter {
+	lineChan := make(chan *Line)
+	lineBuffer := NewLineBuffer(outputChan)
+	delimiter := SingleLineDelimiter{
+		lineChan:   lineChan,
+		lineBuffer: lineBuffer,
+	}
+	go delimiter.run()
+	return &delimiter
+}
+
+// LineChan returns lineChan
+func (d *SingleLineDelimiter) LineChan() chan *Line {
+	return d.lineChan
+}
+
+// run reads lines from lineChan to process them
+func (d *SingleLineDelimiter) run() {
+	for line := range d.lineChan {
+		d.process(line)
+	}
+}
+
+// process sends or truncates lines using lineBuffer
+func (d *SingleLineDelimiter) process(line *Line) {
+	lineLen := len(line.content)
+	if lineLen < contentLenLimit {
+		// send line
+		d.lineBuffer.Add(line)
+		d.lineBuffer.Flush()
+	} else {
+		// truncate content
+		d.lineBuffer.AddAndTruncate(line)
+	}
+}
+
+// outdatedDuration represents the time we want to wait before sending content stored in lineBuffer
+// when mo more line is received
+const outdatedDuration = 1 * time.Second
+
+// MultiLineDelimiter reads lines from lineChan and uses lineBuffer to send them when a new line matches with re or the timer is fired
+type MultiLineDelimiter struct {
+	lineChan   chan *Line
+	lineBuffer *LineBuffer
+	re         *regexp.Regexp
+	timer      *time.Timer
+	mu         sync.Mutex
+}
+
+// NewMultiLineDelimiter returns a new MultiLineDelimiter
+func NewMultiLineDelimiter(outputChan chan *Output, re *regexp.Regexp) *MultiLineDelimiter {
+	lineChan := make(chan *Line)
+	lineBuffer := NewLineBuffer(outputChan)
+	timer := time.NewTimer(outdatedDuration)
+	delimiter := MultiLineDelimiter{
+		lineChan:   lineChan,
+		lineBuffer: lineBuffer,
+		re:         re,
+		timer:      timer,
+	}
+	go delimiter.run()
+	return &delimiter
+}
+
+// LineChan returns lineChan
+func (d *MultiLineDelimiter) LineChan() chan *Line {
+	return d.lineChan
+}
+
+// run sends content when timer expires and reads lines from lineChan to process them
+func (d *MultiLineDelimiter) run() {
+	// send content when timer fires
+	go func() {
+		for range d.timer.C {
+			d.mu.Lock()
+			d.lineBuffer.Flush()
+			d.mu.Unlock()
+		}
+	}()
+
+	// read and process lines
+	for line := range d.lineChan {
+		d.mu.Lock()
+		d.timer.Stop()
+		d.process(line)
+		d.timer.Reset(outdatedDuration)
+		d.mu.Unlock()
+	}
+}
+
+// process handles new lines
+func (d *MultiLineDelimiter) process(line *Line) {
+	if d.re.Match(line.content) {
+		// send content in lineBuffer
+		d.lineBuffer.Flush()
+	}
+	if !d.lineBuffer.IsEmpty() {
+		// add '\n' to content in lineBuffer
+		d.lineBuffer.AddEndOfLine()
+	}
+	if len(line.content) < contentLenLimit {
```

Comment:
should we check `line.content` or `line.content  + lineBuffer.lenght`?

Reply:
We could if it makes more sense but it's not required since I reduced the value of  `contentLenLimit` to 256Mb while it was 1Mb before

Reply:
but what if we were sending many lines all under the limit?